src/compiler/ng-jest-compiler.spec.ts

```diff
@@ -1,34 +1,587 @@
+            expect(code).toContain(`external_1 = require("./external");`);
+            expect(code).toContain(dedent`
+                MyDir.ctorParameters = () => [
+                    { type: external_1.Dep }
+                ];
+                exports.MyDir = MyDir = tslib_1.__decorate([
+                    (0, core_1.Directive)()
+                ], MyDir);
+            `);
+        });
+
+        it('should be able to serialize circular constructor parameter type', () => {
+            const { code, diagnostics } = transformCjs(`
+                import {Directive, Optional, Inject, SkipSelf} from '@angular/core';
+
+                @Directive()
+                export class MyDir {
+                    constructor(@Optional() @SkipSelf() @Inject(MyDir) parentDir: MyDir|null) {}
+                }
+            `);
+
+            expect(diagnostics.length).toBe(0);
+            expect(code).toContain(dedent`
+                let MyDir = class MyDir {
+                    constructor(parentDir) { }
+                };
+                exports.MyDir = MyDir;
+                MyDir.ctorParameters = () => [
+                    { type: MyDir, decorators: [{ type: core_1.Optional }, { type: core_1.SkipSelf }, { type: core_1.Inject, args: [MyDir,] }] }
+                ];
+                exports.MyDir = MyDir = tslib_1.__decorate([
+                    (0, core_1.Directive)()
+                ], MyDir);
+            `);
+        });
+
+        it.failing('should not capture constructor parameter types when not resolving to a value', () => {
+            // This test doesn't pass because `transpileModule` can't resolve `import *`
+            const { code, diagnostics } = transformCjs(`
+                import {Directive, Inject} from '@angular/core';
+                import * as angular from './external';
+                import {IOverlay, KeyCodes} from './external';
+                import TypeFromDefaultImport from './external';
+
+                @Directive()
+                export class MyDir {
+                    constructor(@Inject('$state') param: angular.IState,
+                                @Inject('$overlay') other: IOverlay,
+                                @Inject('$default') fromDefaultImport: TypeFromDefaultImport,
+                                @Inject('$keyCodes') keyCodes: KeyCodes) {}
+                }
+            `);
+
+            expect(diagnostics.length).toBe(0);
+            expect(code).not.toContain('external');
+            expect(code).toContain(dedent`
+                MyDir.ctorParameters = () => [
+                    { type: undefined, decorators: [{ type: core_1.Inject, args: ['$state',] }] },
+                    { type: undefined, decorators: [{ type: core_1.Inject, args: ['$overlay',] }] },
+                    { type: undefined, decorators: [{ type: core_1.Inject, args: ['$default',] }] },
+                    { type: undefined, decorators: [{ type: core_1.Inject, args: ['$keyCodes',] }] }
```

Comment:
If `type` is expected to be `undefined`, then why are we trying to resolve the `import *`? It doesn't seem like we shouldn't need that type? I wonder if the correct behavior on resolution error would just be to fall back to `any` as the type (and emit `undefined`)?

Reply:
This case produces very interesting output
```
MyDir.ctorParameters = () => [
{ type: undefined, decorators: [{ type: core_1.Inject, args: ['$state',] }] },
{ type: external_1.IOverlay, decorators: [{ type: core_1.Inject, args: ['$overlay',] }] },
{ type: external_2.default, decorators: [{ type: core_1.Inject, args: ['$default',] }] },
{ type: external_1.KeyCodes, decorators: [{ type: core_1.Inject, args: ['$keyCodes',] }] }
];
exports.MyDir = MyDir = tslib_1.__decorate([
(0, core_1.Directive)()
], MyDir);
```

Reply:
Hmm, any of these could be an interface which wouldn't be valid in `type` though. I wonder if the type system is falling back to `any` because it can't be resolved and then we're assuming the import is a valid value reference. I wonder if the safer assumption would be to fall back to `type: undefined` in that scenario?

Reply:
Since I’m using a “standalone” `Program` which doesn’t have enough information to validate which imports are interfaces or non interfaces, as well as not able to validate which import paths can be resolved which leads to the situation where the `Program` just assumes those paths exist or those are non interfaces?

Reply:
Here are a few things from my debugging:

### Failed scenarios

- Given file content:
```
import {Directive} from '@angular/core';
import externalFile from './other-file';
import * as foo from './other-file-foo';

@Directive()
export class MyDir {
      constructor(other: externalFile.MyOtherClass,
                           hello: foo.MyFooClass) {}
}
```

- `SymbolObject` shape:
```
symbol SymbolObject {
      flags: 34078720,
      escapedName: 'MyOtherClass',
      declarations: undefined,
      valueDeclaration: undefined,
      id: 0,
      mergeId: 0,
      parent: SymbolObject {
        flags: 34078720,
        escapedName: 'externalFile',
        declarations: undefined,
        valueDeclaration: undefined,
        id: 0,
        mergeId: 0,
        parent: undefined,
        members: undefined,
        exports: undefined,
        exportSymbol: undefined,
        constEnumOnlyModule: undefined,
        isReferenced: undefined,
        lastAssignmentPos: undefined,
        links: SymbolLinks { checkFlags: 1048576, declaredType: [TypeObject] }
      },
      members: undefined,
      exports: undefined,
      exportSymbol: undefined,
      constEnumOnlyModule: undefined,
      isReferenced: undefined,
      lastAssignmentPos: undefined,
      links: SymbolLinks {
        checkFlags: 1048576,
        declaredType: TypeObject {
          flags: 1,
          checker: [Object],
          id: 6,
          intrinsicName: 'unresolved',
          debugIntrinsicName: undefined,
          objectFlags: 52953088
        }
      }
 }
```
with this symbol object, I will always run into this line https://github.com/angular/angular/blob/d108320ea394e827fc23011b54d12e168287bce4/packages/compiler-cli/src/ngtsc/transform/jit/src/downlevel_decorators_transform.ts#L363

### Passed scenario

- Given file content
```
import {Directive} from '@angular/core';
import {MyOtherClass} from './other-file';

@Directive()
export class MyDir {
       constructor(other: MyOtherClass) {}
}
```

- `SymbolObject` shape:
```
symbol <ref *1> SymbolObject {
      flags: 2097152,
      escapedName: 'MyOtherClass',
      declarations: [
        NodeObject {
          pos: 82,
          end: 94,
          kind: 276,
          id: 0,
          flags: 0,
          modifierFlagsCache: 536870912,
          transformFlags: 0,
          parent: [NodeObject],
          original: undefined,
          emitNode: undefined,
          symbol: [Circular *1],
          localSymbol: undefined,
          isTypeOnly: false,
          propertyName: undefined,
          name: [IdentifierObject]
        }
      ],
      valueDeclaration: undefined,
      id: 9,
      mergeId: 0,
      parent: undefined,
      members: undefined,
      exports: undefined,
      exportSymbol: undefined,
      constEnumOnlyModule: undefined,
      isReferenced: 788968,
      lastAssignmentPos: undefined,
      links: undefined
}
```

with this `SymbolObject`, it will never run into this https://github.com/angular/angular/blob/main/packages/compiler-cli/src/ngtsc/transform/jit/src/downlevel_decorators_transform.ts#L363

Reply:
In retrospect, `any` might not necessarily be a problem. Both type references and value references can be typed `any`, so that might be fine. I wonder if it's possible to ask TS whether or not each import is a value reference or a type reference. I'm guessing it's assuming imported symbols are a value reference when it can't resolve the import and we'd probably want to reduce that assumption to a type reference in this scenario.

Unfortunately I'm not very familiar with the TS API and docs are pretty lacking, so I'm not sure about the best way to verify that.

Reply:
For the 2nd case, can we unconditionally *NOT* generating `undefined` so that it’s possible to run tests and leave it up to Jest runtime to validate if that is valid runtime value? 

For example:

`import foo from ‘./foo’` would resolve in `type` as `type: foo.<something>`

And the same applies to `import * as foo from ‘./foo’`

If we set `type: undefined`, it will break all existing tests and it makes it not possible to run tests in transpilation mode because there will be error like `Can't resolve all parameters for AppComponent: (?)` which I observe when debugging this issue https://github.com/thymikee/jest-preset-angular/issues/963